db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OptimizerImpl.java

```diff
@@ -2405,4 +2415,38 @@ protected long getMinTimeout() {
     protected long getMaxTimeout() {
         return Long.MAX_VALUE; // milliseconds
     }
+
+    /**
+     * In the presence of consecutive broadcast joins, we should not only check whether
+     * individual join can fit in memory but also the accumulative memory usage.
+     * The memory check for individual join has already been conducted as part of the broadcast join costing,
+     * here, we check whether the accumulative memory usage is still under the limit.
+     * @return true if the accumulative memory usage is under the limit, false otherwise
+     */
+
+    boolean checkPathMemoryUsage(Optimizable optimizable, boolean checkSortAvoidancePlan) {
+        AccessPath currentAp = optimizable.getCurrentAccessPath();
+        if (currentAp.getJoinStrategy().getJoinStrategyType() != JoinStrategy.JoinStrategyType.BROADCAST)
+            return true;
+
+        double memoryAlreadyConsumed = 0;
+        for (int i=joinPosition-1; i>=0; i--) {
+            AccessPath ap;
+
+            if (checkSortAvoidancePlan)
+                ap = optimizableList.getOptimizable(proposedJoinOrder[i]).getBestSortAvoidancePath();
+            else
+                ap = optimizableList.getOptimizable(proposedJoinOrder[i]).getBestAccessPath();
+
+            if (ap == null || ap.getJoinStrategy().getJoinStrategyType() != JoinStrategy.JoinStrategyType.BROADCAST)
+                break;
+            memoryAlreadyConsumed += ap.getCostEstimate().getBase().getEstimatedHeapSize();
```

Comment:
Is this getting the estimated heap size of the join or the inner table? Because it should only take into account the size of the inner table